homeassistant/components/hyperion/config_flow.py

```diff
@@ -0,0 +1,422 @@
+        #       'sslServer': '8092',
+        #       'protoBuffer': '19445',
+        #       'flatBuffer': '19400'
+        #   },
+        #   'presentationURL': 'index.html',
+        #   'iconList': {
+        #       'icon': {
+        #           'mimetype': 'image/png',
+        #           'height': '100',
+        #           'width': '100',
+        #           'depth': '32',
+        #           'url': 'img/hyperion/ssdp_icon.png'
+        #       }
+        #   },
+        #   'ssdp_usn': 'uuid:f9aab089-f85a-55cf-b7c1-222a72faebe9',
+        #   'ssdp_ext': '',
+        #   'ssdp_server': 'Raspbian GNU/Linux 10 (buster)/10 UPnP/1.0 Hyperion/2.0.0-alpha.8'}
+        data = {}
+
+        data[CONF_HOST] = urlparse(discovery_info[ATTR_SSDP_LOCATION]).hostname
+        try:
+            self._port_ui = urlparse(discovery_info[ATTR_SSDP_LOCATION]).port
+        except ValueError:
+            self._port_ui = const.DEFAULT_PORT_UI
+
+        try:
+            data[CONF_PORT] = int(
+                discovery_info.get("ports", {}).get(
+                    "jsonServer", const.DEFAULT_PORT_JSON
+                )
+            )
+        except ValueError:
+            data[CONF_PORT] = const.DEFAULT_PORT_JSON
+
+        hyperion_id = discovery_info.get(ATTR_UPNP_SERIAL)
+        if hyperion_id:
+            # For discovery mechanisms, we set the unique_id as early as possible to
+            # avoid discovery popping up a duplicate on the screen. The unique_id is set
+            # authoritatively later in the flow by asking the server to confirm its id
+            # (which should theoretically be the same as specified here)
+            await self.async_set_unique_id(hyperion_id)
+            self._abort_if_unique_id_configured()
+        else:
+            return self.async_abort(reason="no_id")
+        return await self.async_step_user(user_input=data)
+
+    async def async_step_user(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by the user."""
+        if not user_input:
+            return self._show_setup_form()
+        self._data = user_input
+
+        # First connect without attempting to login and determine if
+        # authentication is required.
+        async with await self._create_client(raw_connection=True) as hyperion_client:
+            if not hyperion_client:
+                return self._show_setup_form({CONF_BASE: "connection_error"})
+            auth_resp = await hyperion_client.async_is_auth_required()
+
+        # Could not determine if auth is required, show error.
+        if not client.ResponseOK(auth_resp):
+            return self._show_setup_form({CONF_BASE: "auth_required_error"})
+
+        if auth_resp.get(const.KEY_INFO, {}).get(const.KEY_REQUIRED) is True:
+            # Auth is required, show the form to get the token.
+            return self._show_auth_form()
+        return await self._show_confirm_form()
```

Comment:
Don't directly call the auth form function but instead delegate that to the auth step.

```suggestion
            return await self.async_step_auth()
        return await self.async_step_confirm()
```